[UnifiedPDF] Zooming via HUD buttons, keyboard, and context menu

hortont424 opened the pull request:
#### 12f1eb0ad174c46f8acd8586704f9934f17612cb
<pre>
[UnifiedPDF] Zooming via HUD buttons, keyboard, and context menu
<a href="https://bugs.webkit.org/show_bug.cgi?id=267899">https://bugs.webkit.org/show_bug.cgi?id=267899</a>
<a href="https://rdar.apple.com/119632377">rdar://119632377</a>

Reviewed by Simon Fraser.

* Source/WebCore/platform/LocalizedStrings.cpp:
(WebCore::contextMenuItemPDFTwoPagesContinuous):
(WebCore::contextMenuItemPDFAutoSize):
* Source/WebCore/platform/LocalizedStrings.h:
Expose zoomIn, zoomOut, and actualSize strings.

* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::contextMenuItemTagFromDisplayMode const):
(WebKit::UnifiedPDFPlugin::displayModeFromContextMenuItemTag const):
Move these in from the header.

(WebKit::UnifiedPDFPlugin::createContextMenu const):
(WebKit::UnifiedPDFPlugin::performContextMenuAction):
Add context menu items for the three new actions.

(WebKit::UnifiedPDFPlugin::zoomIn):
(WebKit::UnifiedPDFPlugin::zoomOut):
Implement these via setPageScaleFactor(). These back the HUD actions.

(WebKit::UnifiedPDFPlugin::pageScaleFactor const): Deleted.
Don&apos;t implement this getter in terms of PluginView&apos;s cached plugin scale;
just use ours! This way we don&apos;t get out of sync when we call
setPageScaleFactor internally.

* Source/WebKit/WebProcess/Plugins/PluginView.cpp:
(WebKit::PluginView::setPageScaleFactor):
(WebKit::PluginView::pageScaleFactor const):
* Source/WebKit/WebProcess/Plugins/PluginView.h:
Stop storing the plugin scale in PluginView, since it&apos;s duplicative of state
in each of the children. Just defer to them.

* Tools/MiniBrowser/mac/WK2BrowserWindowController.m:
(-[WK2BrowserWindowController resetZoom:]):
Reset page scale in addition to page zoom in MiniBrowser, to match Safari&apos;s behavior.

Canonical link: <a href="https://commits.webkit.org/273342@main">https://commits.webkit.org/273342@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a90d8af1905fe1f0e2947e6d4bd4703ccdd9fbc5

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/35138 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/14067 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/37259 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/37892 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/31722 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/36298 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/16447 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/11136 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/30646 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/35682 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/11901 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/31341 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/10435 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/10491 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/31446 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/39139 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/31964 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/31776 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/36490 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/10610 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/8540 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/34489 "Passed tests") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/12388 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8046 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/11135 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/11450 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

rr-codes commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -64,6 +64,12 @@
 
+// FIXME: We should rationalize these with the values in ViewGestureController.
+// For now, we'll leave them differing as they do in PDFPlugin.
+static const CGFloat minimumZoomScale = 0.2;
```
constexpr?

aprotyas commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -908,29 +907,64 @@
+        || tag == ContextMenuItemTag::SinglePageContinuous
+        || tag == ContextMenuItemTag::TwoPages
+        || tag == ContextMenuItemTag::TwoPagesContinuous);
+
```
Seems redundant with the `default` case

hortont424 replied:
Yeah I don't get it (I didn't add this, just moved and reformatted it). I'll get rid of it.

smfr commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -64,6 +64,12 @@
+// For now, we'll leave them differing as they do in PDFPlugin.
+static const CGFloat minimumZoomScale = 0.2;
+static const CGFloat maximumZoomScale = 6.0;
+static const CGFloat zoomIncrement = 1.18920;
```
Weird

hortont424 replied:
There's a note in PDFKit about how this is weird as well!

aprotyas commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -908,29 +907,64 @@
 
-    // FIXME: We should also set the openInPreviewIndex when UnifiedPdfPlugin::openWithPreview is implemented.
+    auto addSeparator = [&] {
+        menuItems.append({ String(), 0, invalidContextMenuItemTag, ContextMenuItemEnablement::Disabled, ContextMenuItemHasAction::No, ContextMenuItemIsSeparator::Yes });
```
Nit: Use `Vector::constructAndAppend()` instead.

aprotyas replied:
Having said that, this file already has many `Vector::append()` calls. 🤷

hortont424 replied:
Did not know about this, apparently Saam added it in 2016? wild

aprotyas reviewed (commented):
LGTM

hortont424 committed:
[UnifiedPDF] Zooming via HUD buttons, keyboard, and context menu
https://bugs.webkit.org/show_bug.cgi?id=267899
rdar://119632377

Reviewed by Simon Fraser.

* Source/WebCore/platform/LocalizedStrings.cpp:
(WebCore::contextMenuItemPDFTwoPagesContinuous):
(WebCore::contextMenuItemPDFAutoSize):
* Source/WebCore/platform/LocalizedStrings.h:
Expose zoomIn, zoomOut, and actualSize strings.

* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::contextMenuItemTagFromDisplayMode const):
(WebKit::UnifiedPDFPlugin::displayModeFromContextMenuItemTag const):
Move these in from the header.

(WebKit::UnifiedPDFPlugin::createContextMenu const):
(WebKit::UnifiedPDFPlugin::performContextMenuAction):
Add context menu items for the three new actions.

(WebKit::UnifiedPDFPlugin::zoomIn):
(WebKit::UnifiedPDFPlugin::zoomOut):
Implement these via setPageScaleFactor(). These back the HUD actions.

(WebKit::UnifiedPDFPlugin::pageScaleFactor const): Deleted.
Don't implement this getter in terms of PluginView's cached plugin scale;
just use ours! This way we don't get out of sync when we call
setPageScaleFactor internally.

* Source/WebKit/WebProcess/Plugins/PluginView.cpp:
(WebKit::PluginView::setPageScaleFactor):
(WebKit::PluginView::pageScaleFactor const):
* Source/WebKit/WebProcess/Plugins/PluginView.h:
Stop storing the plugin scale in PluginView, since it's duplicative of state
in each of the children. Just defer to them.

* Tools/MiniBrowser/mac/WK2BrowserWindowController.m:
(-[WK2BrowserWindowController resetZoom:]):
Reset page scale in addition to page zoom in MiniBrowser, to match Safari's behavior.

Canonical link: https://commits.webkit.org/273342@main
